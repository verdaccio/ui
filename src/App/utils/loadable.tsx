import React, { lazy, Suspense } from 'react';
export default (importCallback: any) => {
  const TargetComponent = lazy(importCallback);
  return (props: any) => (
    <Suspense fallback={null}>
      <TargetComponent {...props} />
    </Suspense>
  );
};
