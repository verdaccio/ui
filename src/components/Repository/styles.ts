import styled from 'react-emotion';
import Grid from '@material-ui/core/Grid';
import ListItem from '@material-ui/core/ListItem';
import Typography from '@material-ui/core/Typography';

import Github from '../../icons/GitHub';
import colors from '../../utils/styles/colors';

export const Heading = styled(Typography)`
  && {
    font-weight: 700;
    text-transform: capitalize;
  }
`;

export const GridRepo = styled(Grid)`
  && {
    align-items: center;
  }
`;

export const GithubLink = styled('a')`
  && {
    color: ${colors.primary};
  }
`;

export const GithubLogo = styled(Github)`
  && {
    font-size: 40px;
    color: ${colors.primary};
    background-color: ${colors.greySuperLight};
  }
`;

export const RepositoryListItem = styled(ListItem)`
  && {
    padding-left: 0;
    padding-right: 0;
  }
`;
