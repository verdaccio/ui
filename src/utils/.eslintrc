{
    "rules": {
        "no-invalid-this": 0
    }
}