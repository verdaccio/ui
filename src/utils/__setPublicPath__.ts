// @ts-ignore
if (!__DEBUG__) {
  // @ts-ignore
  __webpack_public_path__ = window.VERDACCIO_API_URL.replace(/\/verdaccio\/$/, '/static/'); // eslint-disable-line
}
