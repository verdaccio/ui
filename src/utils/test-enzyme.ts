import { mount, shallow } from 'enzyme';
import React from 'react';

import ThemeProvider from 'verdaccio-ui/design-tokens/ThemeProvider';

const shallowWithTheme = (element: React.ReactElement<any>, ...props: any[]) =>
  shallow(element, {
    wrappingComponent: ThemeProvider,
    ...props,
  });

const mountWithTheme = (element: React.ReactElement<any>, ...props: any[]): any =>
  mount(element, {
    wrappingComponent: ThemeProvider,
    ...props,
  });

export { mountWithTheme as mount, shallowWithTheme as shallow };
