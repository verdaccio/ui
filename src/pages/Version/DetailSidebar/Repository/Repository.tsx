import styled from '@emotion/styled';
import React from 'react';
import { useTranslation } from 'react-i18next';

import Avatar from 'verdaccio-ui/components/Avatar';
import CopyToClipBoard from 'verdaccio-ui/components/CopyToClipBoard';
import Link from 'verdaccio-ui/components/Link';
import List from 'verdaccio-ui/components/List';
import ListItem from 'verdaccio-ui/components/ListItem';
import ListItemText from 'verdaccio-ui/components/ListItemText';
import Text from 'verdaccio-ui/components/Text';
import { Theme } from 'verdaccio-ui/design-tokens/theme';
import { isURL } from 'verdaccio-ui/utils/url';

import { DetailContext } from '../../context';

import git from './img/git.png';

const StyledText = styled(Text)<{ theme?: Theme }>(props => ({
  fontWeight: props.theme && props.theme.fontWeight.bold,
  textTransform: 'capitalize',
}));

const GithubLink = styled(Link)<{ theme?: Theme }>(({ theme }) => ({
  color: theme?.palette.type === 'light' ? theme?.palette.primary.main : theme?.palette.white,
  ':hover': {
    color: theme?.palette.dodgerBlue,
  },
}));

const RepositoryListItem = styled(ListItem)({
  padding: 0,
  ':hover': {
    backgroundColor: 'transparent',
  },
});

const RepositoryListItemText = styled(ListItemText)({
  padding: '0 10px',
  margin: 0,
});

const RepositoryAvatar = styled(Avatar)({
  borderRadius: '0px',
  padding: '0',
  img: {
    backgroundColor: 'transparent',
  },
});

const Repository: React.FC = () => {
  const detailContext = React.useContext(DetailContext);
  const { t } = useTranslation();

  const { packageMeta } = detailContext;

  if (!packageMeta?.latest?.repository?.url || !isURL(packageMeta.latest.repository.url)) {
    return null;
  }

  const { url } = packageMeta.latest.repository;

  const getCorrectRepositoryURL = (): string => {
    if (!url.includes('git+')) {
      return url;
    }

    return url.split('git+')[1];
  };

  const repositoryURL = getCorrectRepositoryURL();

  return (
    <List dense={true} subheader={<StyledText variant="subtitle1">{t('sidebar.repository.title')}</StyledText>}>
      <RepositoryListItem button={true}>
        <RepositoryAvatar src={git} />
        <RepositoryListItemText
          primary={
            <CopyToClipBoard text={repositoryURL}>
              <GithubLink external={true} to={repositoryURL}>
                {repositoryURL}
              </GithubLink>
            </CopyToClipBoard>
          }
        />
      </RepositoryListItem>
    </List>
  );
};

export default Repository;
