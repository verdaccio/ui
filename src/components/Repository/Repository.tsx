/* eslint react/jsx-max-depth: 0 */

import List from '@material-ui/core/List';
import React, { Component, Fragment, ReactElement } from 'react';

import Avatar from '../../muiComponents/Avatar';
import { DetailContextConsumer } from '../../pages/Version';
import { isURL } from '../../utils/url';
import CopyToClipBoard from '../CopyToClipBoard';

import git from './img/git.png';
import { GithubLink, StyledText, RepositoryListItem, RepositoryListItemText } from './styles';

class Repository extends Component {
  public render(): ReactElement<HTMLElement> {
    return (
      <DetailContextConsumer>
        {context => {
          return context && context.packageMeta && this.renderRepository(context.packageMeta);
        }}
      </DetailContextConsumer>
    );
  }

  private renderRepositoryText(url: string): ReactElement<HTMLElement> {
    return (
      <GithubLink href={url} target="_blank">
        {url}
      </GithubLink>
    );
  }

  private renderRepository = packageMeta => {
    const { repository: { url = null } = {} } = packageMeta.latest;

    if (!url || isURL(url) === false) {
      return null;
    }

    return (
      <Fragment>
        <List dense={true} subheader={<StyledText variant="subtitle1">{'Repository'}</StyledText>}>
          <RepositoryListItem button={true}>
            <Avatar src={git} />
            <RepositoryListItemText primary={this.renderContent(url)} />
          </RepositoryListItem>
        </List>
      </Fragment>
    );
  };

  private renderContent(url: string): ReactElement<HTMLElement> {
    return <CopyToClipBoard text={url}>{this.renderRepositoryText(url)}</CopyToClipBoard>;
  }
}

export default Repository;
