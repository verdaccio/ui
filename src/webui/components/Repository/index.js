/* eslint react/jsx-max-depth: 0 */

import React, {Component} from 'react';
import Avatar from '@material-ui/core/Avatar';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';

import { DetailContextConsumer } from '../../pages/version/index';
import CopyToClipBoard from '../CopyToClipBoard';

import { Heading, GithubLink, RepositoryListItem } from './styles';
import git from './img/git.png';
import { isURL } from '../../utils/url';

class Repository extends Component<any, any> {
  render() {
    return (
      <DetailContextConsumer>
        {(context) => {
          return this.renderRepository(context);
        }}
      </DetailContextConsumer>
    );
  };

  renderRepositoryText(url) {
    return (<GithubLink href={url} target={"_blank"}>{url}</GithubLink>);
  }

  renderRepository = ({packageMeta}) => {
    const { 
      repository: {
        url,
      } = {},
    } = packageMeta.latest;
    
    if (!url || isURL(url) === false) {
      return null;
    }

    return (
      <>
        <List dense={true} subheader={<Heading variant={"subheading"}>{'Repository'}</Heading>}>
          <RepositoryListItem>
            <Avatar src={git} />
            <ListItemText primary={this.renderContent(url)} />
          </RepositoryListItem>
        </List>
      </>
    );
  }
  
  renderContent(url) {
    return (
      <CopyToClipBoard text={url}>
        {this.renderRepositoryText(url)}
      </CopyToClipBoard>
    );
  }
}


export default Repository;
