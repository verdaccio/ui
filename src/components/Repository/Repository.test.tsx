import React from 'react';

import { mount } from '../../utils/test-enzyme';

import Repository from './Repository';

jest.mock('./img/git.png', () => '');

const mockPackageMeta: jest.Mock = jest.fn(() => ({
  latest: {
    homepage: 'https://verdaccio.tld',
    bugs: {
      url: 'https://verdaccio.tld/bugs',
    },
    dist: {
      tarball: 'https://verdaccio.tld/download',
    },
  },
}));

jest.mock('../../pages/Version', () => ({
  DetailContextConsumer: component => {
    return component.children({ packageMeta: mockPackageMeta() });
  },
}));

describe('<Repository /> component', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('should render the component in default state', () => {
    const packageMeta = {
      latest: {
        repository: {
          type: 'git',
          url: 'git+https://github.com/verdaccio/ui.git',
        },
      },
    };

    mockPackageMeta.mockImplementation(() => packageMeta);

    const wrapper = mount(<Repository />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render the component in with no repository data', () => {
    const packageMeta = {
      latest: {},
    };

    mockPackageMeta.mockImplementation(() => packageMeta);

    const wrapper = mount(<Repository />);
    expect(wrapper.html()).toEqual('');
  });

  test('should render the component in with invalid url', () => {
    const packageMeta = {
      latest: {
        repository: {
          type: 'git',
          url: 'git://github.com/verdaccio/ui.git',
        },
      },
    };

    mockPackageMeta.mockImplementation(() => packageMeta);

    const wrapper = mount(<Repository />);
    expect(wrapper.html()).toEqual('');
  });
});
