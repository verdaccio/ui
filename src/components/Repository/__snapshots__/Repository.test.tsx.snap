// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repository /> component should load the component in default state 1`] = `null`;
